File: 238-product-of-array-except-self/238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1] * len(nums)
        suffix = [1] * len(nums)
        acc1 = 1
        acc2 = 1
        for i,num in enumerate(nums):
            acc1 *= num
            prefix[i] = acc1
        for i in range(len(nums) -1, -1, -1):
            acc2 *= nums[i]
            suffix[i] = acc2

        res = [0] * len(nums)
        res[0] = suffix[1]
        res[-1] = prefix[-2]
        for i in range(1, len(res) - 1):
            res[i] = prefix[i-1] * suffix[i+1]
        return res

        

File: 238-product-of-array-except-self/238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> prefix;
        vector<int> suffix;
        int curr = 1;
        for (int num: nums) {
            curr *= num;
            prefix.push_back(curr);
        }
        curr = 1;
        for (int i = nums.size() - 1; i >= 0; i--) {
            curr *= nums[i];
            suffix.push_back(curr);
        }
        reverse(suffix.begin(), suffix.end());
        vector<int> result;
        for (int i = 0; i < nums.size(); i++) {
            if (i == 0) result.push_back(suffix[i+1]);
            else if (i == nums.size() - 1) result.push_back(prefix[i-1]);
            else result.push_back(prefix[i-1] * suffix[i+1]);
        }
        return result;

    }
};